cpp/src/parquet/encoding.cc

```diff
@@ -597,7 +757,44 @@ void DictEncoderImpl<ByteArrayType>::Put(const arrow::Array& values) {
 
 template <typename DType>
 void DictEncoderImpl<DType>::PutDictionary(const arrow::Array& values) {
-  ParquetException::NYI(values.type()->ToString());
+  if (this->num_entries() > 0) {
+    throw ParquetException("Can only call PutDictionary on an empty DictEncoder");
+  }
+
+  using ArrayType = typename arrow::CTypeTraits<typename DType::c_type>::ArrayType;
+  const auto& data = checked_cast<const ArrayType&>(values);
+  if (data.null_count() > 0) {
+    throw ParquetException("Inserted dictionary cannot cannot contain nulls");
+  }
+
+  dict_encoded_size_ += static_cast<int>(sizeof(typename DType::c_type) * data.length());
+  for (int64_t i = 0; i < data.length(); i++) {
+    ARROW_IGNORE_EXPR(
+        memo_table_.GetOrInsert(data.Value(i),
+                                /*on_found=*/[](int32_t memo_index) {},
+                                /*on_not_found=*/[](int32_t memo_index) {}));
+  }
+}
+
+template <>
+void DictEncoderImpl<FLBAType>::PutDictionary(const arrow::Array& values) {
+  AssertFixedSizeBinary(values, type_length_);
+  if (this->num_entries() > 0) {
+    throw ParquetException("Can only call PutDictionary on an empty DictEncoder");
+  }
+
+  const auto& data = checked_cast<const arrow::FixedSizeBinaryArray&>(values);
+  if (data.null_count() > 0) {
+    throw ParquetException("Inserted dictionary cannot cannot contain nulls");
+  }
```

Comment:
It's a bit of a bummer to have these assertions duplicated, but it can always be cleaned up later

Reply:
I've cleaned them up a bit now